### PR TITLE
[Serve] Add more nuanced checks for http proxy status errors

### LINKED ISSUE #47895: [Serve] Ray Dashboard reports HTTP 204 under Error QPS
### What happened + What you expected to happen

When a deployment returns an HTTP 204, it is reported as an error in the Ray Dashboard. Specifically, these queries are reported under the "Error QPS" charts. An HTTP 2xx should not be considered as an error.

Screenshot 2024-10-04 at 9 07 06 AM

### Versions / Dependencies

macOS 14.3
Python: 3.12
ray[serve] 2.37.0

### Reproduction script

```python
from ray import serve
from fastapi import FastAPI, Response


app = FastAPI()


@serve.deployment
@serve.ingress(app)
class Deployment:
    @app.post("/")
    def post(self, text: str) -> str:
        return Response(status_code=204)


deployment_app = Deployment.bind()
```

### Issue Severity

None

### PR DESCRIPTION
**Pull request by Superskyyy**
Closes: https://github.com/ray-project/ray/issues/47895

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/proxy.py [hunk 1]
```diff
@@ -1022,11 +1021,17 @@ async def send_request_to_replica(
...
+
+                        # Check based on standard WebSocket status codes
+                        if status_code in ["1000", "1001"]:
+                            # Normal closure or going away, no error
+                            is_error = False
+                        else:
+                            # Other 1xxx codes are specified as errors
+                            is_error = status_code.startswith("1")
+
+                        status = ResponseStatus(code=status_code, mis_error=is_error)
```
Comment (edoakes) [inline python/ray/serve/_private/proxy.py:1034]:
```suggestion
                        status = ResponseStatus(code=status_code, is_error=is_error)
```

**Review by edoakes — COMMENTED** [review]
Thanks for the contribution!

Could you please also add an e2e test for this behavior to `test_metrics.py`? You should be able to follow other tests there as a blueprint.

**Comment by Superskyyy** [comment]
> Thanks for the contribution!
> 
> Could you please also add an e2e test for this behavior to `test_metrics.py`? You should be able to follow other tests there as a blueprint.

Certainly, I will do that. Thanks

**Comment by edoakes** [comment]
Hey @Superskyyy just checking in, let me know if you need a pointer on the test

**Review by edoakes — APPROVED** [review]

**Pull request merged**